### DISCUSSION TITLE
OpenSSL 3.X - convert engine to provider for cURL TLS authentication

### DISCUSSION THREAD

**Discussion by MscVZoltan** [openssl/openssl#22839] [Q&A]
Hi!

Out developer team is trying to replace an engine to a provider and is a bit stuck now.

We are developing C++ a software, that connects with TLS using cURL and may use a PKCS#11 token for authentication in some cases. We previously made an engine and implemented RSA_METHOD-s rsa_sign (and rsa_verify, which is just using RSA_verify, furthermore it is not called during TLS handshake anymore after the OpenSSL update) and nothing else.

If I understand correctly, now we have to create a provider for OSSL_OP_SIGNATURE and for OSSL_OP_KEYMGMT and implement nearly all functions for it, because how cURL uses the SSL functions, it relies on them heavily: digest sign init/update/final, verify init/update/final, ctx creation and duplication, getting and setting ctx params, giving back a table of possible ctx params, etc.

Is there any other way to achieve our goal, if we only want to implement the signing functionality and nothing else? I know, that the basic concept is that a provider can uniquely implement how the keys should be used, but we just want to somehow "hook" the signing functionality of the default provider and leave the rest as is. Another, not very comfortable option is that we can reuse nearly all code from providers\implementations\signature\rsa_sig and providers\implementations\signature\ecdsa_sig.

**Comment by dingyaoyang** [comment]
mark. I have same problem. I‘m a novice, and there is a little odd when I imple a ec_key_meth, my sign and verify not be called. I don't know if it's my problem or not.
Waiting for a solution.

**Comment by tanjahummel** [comment]
Hi, I have a similar question.
We are also developing C++ software that uses TLS with an own signature implementation and are trying to convert the engine to a provider.

Currently, the engine is set at the PKEY object which is passed to the SSL ctx like this:
EVP_PKEY* pub_key = X509_get_pubkey(certificate);
EVP_PKEY_set1_engine(pub_key, engine);
SSL_CTX_use_PrivateKey(ctx, pub_key);

I didn't found a similar way to configure the PKEY object to use a specific keymgmt or signature provider. How should this be done with the provider concept?

**Comment by mattcaswell** [comment]
To implement signing in a provider for TLS you only need to implement a subset of the functions for OSSL_OP_SIGNATURE and for OSSL_OP_KEYMGMT. For example, you don't need to implement any of the "verify" functions if you're only interested in signing, and you don't need to implement any of the "key gen" functions if you're not interested in that.

I would recommend setting things up in the following way:

1) Do *not* implement support for exporting of *private* key components in your key manager

2) Ensure the default provider and your provider are both loaded into your libctx

3) Set your global default properties to *not* prefer your provider. The "?" below indicates preference, i.e. prefer any other provider but yours. This way your provider will only be used when it absolutely has to be.

```` c
EVP_set_default_properties(NULL, "?provider!=myprov");
````
4) Create your private key and associate it with your provider. For example one way might be like this (not tested - error handling omitted for brevity):

```` c
EVP_PKEY_CTX *ctx= EVP_PKEY_CTX_new_from_name(NULL, "RSA", "provider=myprov");
EVP_PKEY *pkey = NULL;
OSSL_PARAM keyparams[] = {...whatever...};

EVP_PKEY_fromdata_init(ctx);
EVP_PKEY_fromdata(ctx, &pkey, EVP_PKEY_KEYPAIR, keyparams);
EVP_PKEY_CTX_free(ctx);
````
5) Create your SSL_CTX and set the private key you just created in it in the normal way.

When a TLS connection is created it will always try and use the default provider whenever it can. When it comes to signing anything with your private key, the key is only associated with your provider and cannot be exported from your provider - so OpenSSL will perform the signing operation only using your provider.

**Comment by mattcaswell** [comment]
If you haven't already done so watch our various talks about writing a provider. Most relevant here are:

https://www.youtube.com/watch?v=ztfbzsnHqKU
https://www.youtube.com/watch?v=8ccrNMieRrA
